File: doc/examples/fetch_catalog_overlap.js
#!/usr/bin/env node
// Resolve the distinct options + a recommended default for a language from
// catalog-overlap.json. No dependencies - uses the built-in fetch()
// (Node 18+, or any browser).
//
// No `default` field is published (removed 2026-07-28) - it's a one-line
// computation from `priority` + `ids` that every client can do itself, so
// this example does exactly that rather than relying on a precomputed field.
//
// Usage: node fetch_catalog_overlap.js <iso> [canon]

const URL = "https://cdn.bibel.wiki/dbt/_app/catalog-overlap.json";
const SOURCE_NAME = { d: "dbt", h: "helloao", p: "pkf" };

function pickDefault(ids, priority) {
  const rank = Object.fromEntries(priority.map((name, i) => [name, i]));
  return ids.reduce((best, id) => {
    const source = SOURCE_NAME[id.split(":")[0]];
    const bestSource = SOURCE_NAME[best.split(":")[0]];
    return rank[source] < rank[bestSource] ? id : best;
  });
}

async function main() {
  const iso = process.argv[2];
  const canon = process.argv[3] || "nt";
  if (!iso) {
    console.error("usage: node fetch_catalog_overlap.js <iso> [canon]");
    process.exit(1);
  }

  const data = await (await fetch(URL)).json();
  const clusters = data.entries[`${iso}:${canon}`];

  if (!clusters) {
    console.log(
      `No comparison data for '${iso}' (${canon}). Either only one candidate exists ` +
      `for this language (nothing to compare against - check catalog-index.json), ` +
      `or nothing has been fetched yet.`
    );
    return;
  }

  console.log(`Distinct options for '${iso}' (${canon}):`);
  for (const cluster of clusters) {
    const { ids } = cluster;
    if (cluster.r === false) {
      const note = cluster.confirmed_removed ? "[CONFIRMED REMOVED]" : "[currently unreachable]";
      console.log(`  ${ids[0]} ${note}`);
      continue;
    }
    const def = pickDefault(ids, data.priority);
    const alternatives = ids.filter((i) => i !== def);
    let line = `  -> ${def}`;
    if (alternatives.length) line += `  (identical to: ${alternatives.join(", ")})`;
    console.log(line);
    if (cluster.likely) {
      console.log(
        `     not identical to anything - closest is ${cluster.closest} ` +
        `(${cluster.likely}, score ${cluster.score})`
      );
    }
  }
}

main();

File: doc/examples/fetch_catalog_overlap.py
#!/usr/bin/env python3
"""Resolve the distinct options + a recommended default for a language from
catalog-overlap.json. No dependencies beyond the Python standard library.

No `default` field is published (removed 2026-07-28) - it's a one-line
computation from `priority` + `ids` that every client can do itself, so
this example does exactly that rather than relying on a precomputed field.

Usage:
    python3 fetch_catalog_overlap.py <iso> [canon]
    python3 fetch_catalog_overlap.py spa nt
"""
import json
import sys
import urllib.request

URL = "https://cdn.bibel.wiki/dbt/_app/catalog-overlap.json"
SOURCE_NAME = {"d": "dbt", "h": "helloao", "p": "pkf"}


def pick_default(ids, priority):
    rank = {name: i for i, name in enumerate(priority)}
    return min(ids, key=lambda i: rank[SOURCE_NAME[i.split(":", 1)[0]]])


def main():
    if len(sys.argv) < 2:
        sys.exit("usage: fetch_catalog_overlap.py <iso> [canon]")
    iso = sys.argv[1]
    canon = sys.argv[2] if len(sys.argv) > 2 else "nt"

    # Cloudflare (fronting cdn.bibel.wiki) rejects urllib's default
    # User-Agent with a 403 - always set a real one.
    req = urllib.request.Request(URL, headers={"User-Agent": "bibles-examples/1.0"})
    with urllib.request.urlopen(req) as r:
        data = json.loads(r.read())

    clusters = data["entries"].get(f"{iso}:{canon}")
    if not clusters:
        print(f"No comparison data for '{iso}' ({canon}). Either only one candidate exists "
              f"for this language (nothing to compare against - check catalog-index.json), "
              f"or nothing has been fetched yet.")
        return

    print(f"Distinct options for '{iso}' ({canon}):")
    for cluster in clusters:
        ids = cluster["ids"]
        if cluster.get("r") is False:
            note = " [CONFIRMED REMOVED]" if cluster.get("confirmed_removed") else " [currently unreachable]"
            print(f"  {ids[0]}{note}")
            continue
        default = pick_default(ids, data["priority"])
        alternatives = [i for i in ids if i != default]
        line = f"  -> {default}"
        if alternatives:
            line += f"  (identical to: {', '.join(alternatives)})"
        print(line)
        if "likely" in cluster:
            print(f"     not identical to anything - closest is {cluster['closest']} "
                  f"({cluster['likely']}, score {cluster['score']})")


if __name__ == "__main__":
    main()
